### PR TITLE
[SYCL] Fix working with subbuffers with non-nil offset

### PR DESCRIPTION
**Pull request by s-kanaev**

### CODE REVIEW

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 1]
```diff
@@ -320,11 +320,13 @@ Command *Scheduler::GraphBuilder::addHostAccessor(Requirement *Req) {
 
   AllocaCommandBase *SrcAllocaCmd = nullptr;
 
-  if (Record->MAllocaCommands.empty())
+  if (Record->MAllocaCommands.empty() || IsSuitableSubReq(Req))
     SrcAllocaCmd = getOrCreateAllocaForReq(Record, Req, HostQueue);
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/graph_builder.cpp]: getOrCreateAkkicaForReq() method contains a proper procedure to create/find alloca cmd for subbuffer. That is why there is this kind of hack: regardless of whether alloca commands collection is non-empty go on and getOrCreate it. The method will lookup for it in first place, though. It'll not find the correct alloca command 'cause there is no equality of MOffsetInBytes. After that the method will create a proper alloca cmd for subbuffer.

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 2]
```diff
@@ -320,11 +320,13 @@ Command *Scheduler::GraphBuilder::addHostAccessor(Requirement *Req) {
 
   AllocaCommandBase *SrcAllocaCmd = nullptr;
 
-  if (Record->MAllocaCommands.empty())
+  if (Record->MAllocaCommands.empty() || IsSuitableSubReq(Req))
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/graph_builder.cpp]: getOrCreateAllocaForReq() method contains a proper procedure to create/find alloca cmd for subbuffer. That is why there is this kind of hack: regardless of whether alloca commands collection is non-empty go on and getOrCreate it. The method will lookup for it in first place, though. It'll not find the correct alloca command 'cause there is no equality of MOffsetInBytes. After that the method will create a proper alloca cmd for subbuffer.

Comment (s-kanaev) [inline sycl/source/detail/scheduler/graph_builder.cpp]: Another approach here might be to check if SrcAllocaCmd is null and create it the same way getOrCreateAllocaForReq does if IsSuitableSubReq(Req) returns true. Thing is, this approach would violate code reuse.

Comment (s-kanaev) [reply sycl/source/detail/scheduler/graph_builder.cpp]: The second commit implements this second approach.

**Comment by s-kanaev** [comment]
> Suggest doing something like this instead:


Done

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 3]
```diff
@@ -318,14 +318,11 @@ Command *Scheduler::GraphBuilder::addHostAccessor(Requirement *Req) {
     printGraphAsDot("before_addHostAccessor");
   markModifiedIfWrite(Record, Req);
 
-  AllocaCommandBase *SrcAllocaCmd = nullptr;
-
-  if (Record->MAllocaCommands.empty())
-    SrcAllocaCmd = getOrCreateAllocaForReq(Record, Req, HostQueue);
-  else
-    SrcAllocaCmd = findAllocaForReq(Record, Req, Record->MCurContext);
+  AllocaCommandBase *SrcAllocaCmd =
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/graph_builder.cpp]:
```suggestion
  AllocaCommandBase *HostAllocaCmd =
```
Actually it's not always source allocation command.

**Pull request merged**